amm/drink-tests/src/stable_swap_tests.rs

```diff
@@ -0,0 +1,325 @@
+use crate::factory_contract;
+use crate::pair_contract;
+use crate::pair_contract::Pair;
+use crate::router_contract;
+use crate::stable_pool_contract::{self, StablePoolError};
+use crate::utils::*;
+
+use router_contract::Router as _;
+
+use drink::frame_support::sp_runtime::traits::IntegerSquareRoot;
+use drink::frame_support::sp_runtime::traits::Scale;
+use drink::{self, runtime::MinimalRuntime, session::Session};
+use ink_primitives::AccountId;
+use ink_wrapper_types::{Connection, ToAccountId};
+
+const FEE_BPS_DENOM: u128 = 10_000;
+
+fn setup_stable_swap(
+    session: &mut Session<MinimalRuntime>,
+    token_decimals: [u8; 2],
+    token_supply: [u128; 2],
+    amp_coef: u128,
+    _fee_bps: u128,
+    _admin_fee_bps: u128,
+) -> (AccountId, AccountId, AccountId) {
+    // (stable_pool, token_0, token_1)
+    let _ = session.set_actor(BOB);
+    upload_all(session);
+    session
+        .upload_code(stable_pool_contract::upload())
+        .expect("Upload stable_pair_contract code");
+
+    // instantiate tokens
+    let token_0 = psp22_utils::setup_with_amounts(
+        session,
+        "token_0".to_string(),
+        token_decimals[0],
+        token_supply[0],
+        BOB,
+    );
+    let token_1 = psp22_utils::setup_with_amounts(
+        session,
+        "token_1".to_string(),
+        token_decimals[1],
+        token_supply[1],
+        BOB,
+    );
+
+    // instantiate stable_swap
+    let instance = stable_pool_contract::Instance::new_stable(
+        vec![token_0.into(), token_1.into()],
+        token_decimals.to_vec(),
+        amp_coef,
+        // Fees {fee_bps, admin_fee_bps},
+        bob(),
+        Some(charlie()), // fee receiver
+    );
+
+    let stable_swap: stable_pool_contract::Instance = session
+        .instantiate(instance)
+        .unwrap()
+        .result
+        .to_account_id()
+        .into();
+
+    // setup max allowance for stable swap contract on both tokens
+    for token in [token_0, token_1] {
+        psp22_utils::increase_allowance(session, token.into(), stable_swap.into(), u128::MAX, BOB)
+            .unwrap();
+    }
+
+    (stable_swap.into(), token_0.into(), token_1.into())
+}
+
+/// Tests swap of token at index 0 to token at index 1.
+fn setup_test_swap_exact_in(
+    session: &mut Session<MinimalRuntime>,
+    token_decimals: [u8; 2],
+    initial_reserves: [u128; 2],
+    amp_coef: u128,
+    fee_bps: u128,
+    admin_fee_bps: u128,
+    swap_amount_in: u128,
+    expected_swap_amount_out_total_result: Result<u128, StablePoolError>,
+) {
+    let initial_supply = [initial_reserves[0] + swap_amount_in, initial_reserves[1]];
+    let (stable_swap, token_0, token_1) = setup_stable_swap(
+        session,
+        token_decimals,
+        initial_supply,
+        amp_coef,
+        fee_bps,
+        admin_fee_bps,
+    );
+    _ = stable_swap::add_liquidity(
+        session,
+        stable_swap,
+        BOB,
+        1,
+        initial_reserves.to_vec(),
+        bob(),
+    );
```

Comment:
This returns `ContractResult` – let's `unwrap` etc etc just for the sake of consistency.